src/brod_client.erl

```diff
@@ -159,67 +184,36 @@ get_partitions(Client, Topic) ->
       {error, Reason}
   end.
 
-%% @doc Register self() as a partition producer the pid is registered in an ETS
+%% @doc Register self() as a partition producer. The pid is registered in an ETS
 %% table, then the callers may lookup a producer pid from the table and make
 %% produce requests to the producer process directly.
 %% @end
 -spec register_producer(client(), topic(), partition()) -> ok.
 register_producer(Client, Topic, Partition) ->
   Producer = self(),
-  gen_server:cast(Client, {register_producer, Topic, Partition, Producer}).
+  Key = ?PRODUCER_KEY(Topic, Partition),
+  gen_server:cast(Client, {register, Key, Producer}).
 
-%% @doc Get producer of a given topic-partition.
--spec get_producer(client(), topic(), partition()) ->
-        {ok, pid()} | {error, Reason}
-          when Reason :: client_down
-                       | {producer_down, noproc}
-                       | {producer_not_found, topic()}
-                       | {producer_not_found, topic(), partition()}.
-get_producer(ClientPid, Topic, Partition) when is_pid(ClientPid) ->
-  case erlang:process_info(ClientPid, registered_name) of
-    {registered_name, ClientId} ->
-      get_producer(ClientId, Topic, Partition);
-    Err when Err =:= [] orelse Err =:= ?undef ->
-      {error, client_down}
-  end;
-get_producer(ClientId, Topic, Partition) when is_atom(ClientId) ->
-  try
-    case ets:lookup(?ETS(ClientId), ?PRODUCER_KEY(Topic, Partition)) of
-      [] ->
-        %% not yet registered, 2 possible reasons:
-        %% 1. caller is too fast, producers are starting up
-        %%    make a synced call all the way down the supervision tree
-        %%    to the partition producer sup should resolve the race
-        %% 2. bad argument, no such worker started, supervisors should know
-        find_producer(ClientId, Topic, Partition);
-      [?PRODUCER(Topic, Partition, Pid)] ->
-        {ok, Pid}
-    end
-  catch error:badarg ->
-    {error, client_down}
-  end.
+%% @doc Register self() as a partition consumer. The pid is registered in an ETS
+%% table, then the callers may lookup a consumer pid from the table ane make
+%% subscribe calls to the process directly.
+%% @end
+register_consumer(Client, Topic, Partition) ->
+  Consumer = self(),
+  Key = ?CONSUMER_KEY(Topic, Partition),
+  gen_server:cast(Client, {register, Key, Consumer}).
```

Comment:
cast should always work (sooner or later)
unless client process restart --- but, if that's the case, all producers and consumers should be restarted too.